lib/puppet/pops/lookup.rb

```diff
@@ -56,6 +58,31 @@ def self.lookup(name, value_type, default_value, has_default, merge, lookup_invo
         lookup_invocation.emit_debug_info(debug_preamble(names)) if Puppet[:debug]
         fail_lookup(names)
       end
+    else
+      opt_key = Puppet::Pops::Lookup::LookupKey.new(result_with_name[0])
```

Comment:
Placing this code here introduces unnecessary overhead. The task of finding the lookup_options for the key has already been performed once by the `LookupAdapter`. Not only must the lookup key be recreated again here, it's also necessary to retrieve the lookup_options again.

Can this be moved into the adapter where the `LookupKey` is created and the `lookup_options` have been picked up? AFAICT it would only require a minor refactoring of how the merge option is retrieved.